### PR TITLE
Multi signals

### LINKED ISSUE #10: Finish the covidcast_signals function
It's currently in the R package, but not exported. Remaining issues:

* [x] How should the sources/signals be specified? Currently it accepts a data frame, but Ryan thinks separate vectors with recycling would be convenient, so you could specify things like `data_source = "jhu-csse"` , and `signal = c("confirmed_incidence_num", "confirmed_incidence_prop")`.
* [x] Should it be possible to select signals with different `geo_types` in one call?
* [x] How should `plot.covidcast_signal()` handle data frames containing multiple signals? It could make a grid of plots, perhaps; faceting doesn't work easily because different signals are on different scales.

I also think we should provide a convenience function that wraps `pivot_wider` to convert a data frame with multiple signals to have one row per date and one column per signal, named after the signal, instead of one row per observation per signal. People might like such a data frame (it'd be useful for making feature matrices for models).

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1198968030437421/1198970310857497) by [Unito](https://www.unito.io/learn-more)

### PR DESCRIPTION
**Pull request by ryantibs**
Hi all, particularly capnrefsmmat @chinandrew, this finishes off the `covidcast_signals()` function, and provides related signal aggregation functionality.

* Closes #10 and #48.
* Discussion of the design decisions for signal aggregation can be found in the comments on #10. 
* I also did a bunch of clean-up work, like strike direction from the record everywhere (documentation, `covidcast_signal` object, plotting functionality, etc.), which I suggested we should do in #131.

Because I branched this off of evalcast-review (which was branched off of evalcast), the commit log here is huge.  My first "real" commit for this branch starts at b42b353.  Once evalcast gets merged into main #85 (should happen within a couple days), the commit log here will look a lot cleaner, and we can merge this branch.  chinandrew has already provided an informal (?) review, maybe he can take one more look.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1198968030437421/1198970430348989) by [Unito](https://www.unito.io/learn-more)

### CODE REVIEW

**Comment by chinandrew** [comment]
Changed the based of this PR to evalcast so 1) if it gets merged it doesn't accidently merge in all of evalcast to main, and 2) makes the diff and commit log a bit easier to read. It's easy to switch back if needed
Screenshot from 2020-10-28 11-23-29

File: R-packages/covidcast/R/wrangle.R [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+#'   `covidcast_signal` data frames, and `dt` is either a vector of shifts---to
+#'   apply the same shifts for each signal in `x`, or a list of vector of
+#'   shifts---to apply different shifts for each signal in `x`.
+#'
+#' @return Data frame of aggregated signals in "wide" or "long" form, depending 
+#'   on `format`. In "long" form, an extra column `dt` is appended to indicate
+#'   the value of the time-shift. In "wide" form, only the latest issue of data
+#'   is retained, and the columns `data_source`, `signal`, `issue`, `lag`,
+#'   `stderr`, `sample_size` are all dropped from the returned data frame. Each
+#'   unique signal---defined by a combination of data source name, signal name,
```
Comment (chinandrew) [inline R-packages/covidcast/R/wrangle.R:128]: Might be worth noting that these are full joins on everything in the list and not some other join type

**Comment by chinandrew** [comment]
As mentioned in #10 the interface here looks good to me, though someone more familiar with R paradigms should have the final say.

File: R-packages/covidcast/R/wrangle.R (hunk 1 above)
Comment (ryantibs) [reply R-packages/covidcast/R/wrangle.R]: Thanks, done.

**Comment by ryantibs** [comment]
@chinandrew @capnrefsmmat Few more updates.  First, I filled out the vignette so this branch should be good to go.  Andrew approves but suggests somebody more familiar with R takes a look.  To repeat this doesn't need to be merged until evalcast gets merged, which shouldn't be for a few days at least.  If Alex you're too busy then maybe @JedGrabman can take a look.

Second, a spotted a discrepancy in the way I defined `dt` in `aggregate_signals()` and `covidcast_cor()`.  The definition in the former function is more natural; the latter has it flipped.  So I changed everything surrounding `covidcast_cor()` to match `aggregate_signals()` and looked around for code that would break; all I could find was the correlations vignette and the intro-api talk, both of which I fixed.

To summarize: 
- This should be merged into main after evalcast gets merged.
- This requires us to rebuild the documentation and the vignettes.
- Maybe we should record as an update that `covidcast_cor()` now interprets `dt_x` and `dt_y` in the *opposite* way it did before, so that negative values mean lagging and positive values mean leading (which is more natural).

**Comment by capnrefsmmat** [comment]
> Maybe we should record as an update that covidcast_cor() now interprets dt_x and dt_y in the opposite way it did before, so that negative values mean lagging and positive values mean leading (which is more natural).

That can go into NEWS.md, and we can bump the minor version of the package accordingly. Note that this branch doesn't contain #148, so there'll be a conflict if you do this.

**Comment by ryantibs** [comment]
@capnrefsmmat @chinandrew Upon reflection there was really no reason for me to have branched this off of the evalcast branch.  I was just working on something else that I needed the evalcast package for, and then I started implementing these multi signal functions because I also needed them.  But my "real" commits starting with b42b353 have noting to do with evalcast.  

Since the evalcast branch is actually held up with more development needed, what's the best way to jut pick off the commits starting with b42b353 through the end, and move them to a separate branch, which is created off of main?  Then we could just merge that new into main.

**Comment by capnrefsmmat** [comment]
We'd have to carefully cherry-pick, since there *are* changes to evalcast in this series of commits, namely in 
1e674d8. Those already seem to have a merge conflict, so we'd want to cherry-pick every other commit into a new branch.

That's doable, although whether or not we do it, I think it still makes sense to merge evalcast and continue further development of it in new PRs targeting specific features/fixes.

**Comment by chinandrew** [comment]
You should either be able to do a rebase like `git rebase --onto main evalcast multi-signals`, or creating a new branch and cherry-picking the commit range you want: `git cherry-pick start_commit^..end_commit`

**Comment by ryantibs** [comment]
Thanks both of you! I guess I could try to cherry-pick but it might be easiest just to merge evalcast to main as it stands now (Alex I just left a message agreeing with your logic on #85).

**Comment by ryantibs** [comment]
@capnrefsmmat Well I snuck in a few other really small light housekeeping things.  See the details of my last commit.  I managed to do it before you merge this into main (but apparently not before you merged main into this!).

File: R-packages/covidcast/R/cor.R [hunk 2]
```diff
@@ -1,13 +1,18 @@
...
-#'   before computing correlations. These must both be nonnegative, and only one
+#'   before computing correlations. Negative shifts translate into in a lag
+#'   value and positive shifts into a lead value; for example, if `dt = -1`,
+#'   then the value on June 2 that gets reported is the original value on June
+#'   1; if `dt = 0`, then the values are left as is.
+#'
+#' These must both be nonnegative, and only one
 #'   can be positive. If `dt_x = 1`, for example, then data for `x` is shifted
 #'   forward 1 day in time (so, data on June 1 becomes data on June 2, and so
 #'   on). Default is 0 for both.
```
Comment (capnrefsmmat) [inline R-packages/covidcast/R/cor.R:19]: Is this documentation accurate anymore? It still says both must be nonnegative.

File: R-packages/covidcast/R/wrangle.R [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    dplyr::relocate(dt, .before = value)
+
+  # Change class and return
+  class(x) = c("covicast_signal_long", "data.frame")
+  return(x)
+}
+
+#' Pivot covidcast_signal_long object to "wide" format
+#
+#' @export
```
Comment (capnrefsmmat) [inline R-packages/covidcast/R/wrangle.R:270]: Is this sufficient documentation for this function (and its counterpart `covidcast_longer`)?

**Review by capnrefsmmat — CHANGES_REQUESTED** [review]
Noticed two issues while preparing to merge

File: R-packages/covidcast/R/cor.R (hunk 2 above)
Comment (ryantibs) [reply R-packages/covidcast/R/cor.R:19]: Oops that was a paragraph that was intended to be deleted. Good catch, thanks.

File: R-packages/covidcast/R/wrangle.R (hunk 3 above)
Comment (ryantibs) [reply R-packages/covidcast/R/wrangle.R:270]: Thanks, I added a bit more.  This doesn't need extensive documentation because it just acts as if `aggregate_signals()` had been called with a different option, which I now explained.
